spack_environments/build_lbann.sh

```diff
@@ -0,0 +1,247 @@
+#!/bin/sh
+
+SPACK_VERSION=$(spack --version)
+MIN_SPACK_VERSION=0.13.3
+
+compare_versions()
+{
+    local v1=( $(echo "$1" | tr '.' ' ') )
+    local v2=( $(echo "$2" | tr '.' ' ') )
+    local len=$(( ${#v1[*]} > ${#v2[*]} ? ${#v1[*]} : ${#v2[*]} ))
+    #local len=3
+    for ((i=0; i<len; i++))
+    do
+        [ "${v1[i]:-0}" -gt "${v2[i]:-0}" ] && return 1
+        [ "${v1[i]:-0}" -lt "${v2[i]:-0}" ] && return 2
+    done
+    return 0
+}
+
+compare_versions ${SPACK_VERSION} ${MIN_SPACK_VERSION}
+VALID_SPACK=$?
+
+if [[ ${VALID_SPACK} -eq 2 ]]; then
+    echo "Newer version of Spack required.  Detected version ${SPACK_VERSION} requires at least ${MIN_SPACK_VERSION}"
+    exit 1
+fi
+
+SPACK_ENV_DIR=$(dirname ${BASH_SOURCE})
+
+if [[ ${SPACK_ENV_DIR} = "." ]]; then
+    LBANN_HOME=$(dirname ${PWD})
+    SPACK_ENV_DIR=${LBANN_HOME}/spack_environments
+elif [[ ${SPACK_ENV_DIR} = /* ]]; then
+    LBANN_HOME=$(dirname ${SPACK_ENV_DIR})
+else
+    LBANN_HOME=$(dirname ${PWD}/${SPACK_ENV_DIR})
+    SPACK_ENV_DIR=${LBANN_HOME}/spack_environments
+fi
+
+# Detect system parameters
+CLUSTER=$(hostname | sed 's/\([a-zA-Z][a-zA-Z]*\)[0-9]*/\1/g')
+ARCH=$(uname -m)
+SYS=$(uname -s)
+
+# Identify the center that we are running at
+CENTER=
+if [[ ${SYS} = "Darwin" ]]; then
+    CENTER="osx"
+else
+    CORI=$([[ $(hostname) =~ (cori|cgpu) ]] && echo 1 || echo 0)
+    if [[ ${CORI} -eq 1 ]]; then
+	CENTER="nersc"
+	# Make sure to purge and setup the modules properly prior to finding the Spack architecture
+	source ${SPACK_ENV_DIR}/${CENTER}/setup_modules.sh
+    else
+	CENTER="llnl_lc"
+    fi
+fi
+
+SPACK_ARCH=$(spack arch)
+
+SCRIPT=$(basename ${BASH_SOURCE})
+BUILD_DIR=${LBANN_HOME}/build/spack
```

Comment:
I think it would be nicer to have a separate build directory for each system so that different builds don't clobber each other. Something like `lbann/build/gnu.Release.lassen.gov/build`.

Reply:
Done.